src/main/utility/status_bar.rs

```diff
@@ -0,0 +1,190 @@
+                //   differently. For example, when using '{next_line}' some terminals will
+                //   allow the cursor to move outside of the scroll region, and others don't.
+                // {save_cursor}\n{restore_cursor}
+                //   This will scroll the buffer up only if the cursor is on the last row.
+                // {next_line}{prev_line}
+                //   This will move the cursor up only if the cursor is on the last row (to
+                //   match the previous scroll behaviour).
+                // {save_cursor}
+                //   The cursor is currently at the correct location, so save it for later.
+                // \u{1B}[1;{scroll_height}r
+                //   Set the scroll region to include all rows but the last.
+                // {last_line}{msg}{clear}
+                //   Move to the last row and write the message.
+                // {restore_cursor}
+                //   Restore the cursor position.
+                eprint!(
+                    "{save_cursor}{restore_scroll_region}{restore_cursor}\
+                     {save_cursor}\n{restore_cursor}\
+                     {next_line}{prev_line}\
+                     {save_cursor}\
+                     \u{1B}[1;{scroll_height}r\
+                     {last_line}{msg}{clear}\
+                     {restore_cursor}",
+                    save_cursor = SAVE_CURSOR,
+                    restore_scroll_region = RESTORE_SCROLL_REGION,
+                    restore_cursor = RESTORE_CURSOR,
+                    next_line = NEXT_LINE,
+                    prev_line = PREV_LINE,
+                    scroll_height = rows - 1,
+                    last_line = LAST_LINE,
+                    msg = *state.read().unwrap(),
+                    clear = CLEAR,
+                );
+                let _ = std::io::stderr().flush();
+            }
+            std::thread::sleep(redraw_interval);
+        }
+
+        eprint!(
+            "{save_cursor}{last_line}{clear}{restore_scroll_region}{restore_cursor}",
+            save_cursor = SAVE_CURSOR,
+            last_line = LAST_LINE,
+            clear = CLEAR,
+            restore_scroll_region = RESTORE_SCROLL_REGION,
+            restore_cursor = RESTORE_CURSOR,
+        );
+        let _ = std::io::stderr().flush();
+    }
+
+    /// Stop and remove the status bar.
+    pub fn stop(&mut self) {
+        self.stop_flag
+            .swap(true, std::sync::atomic::Ordering::Relaxed);
+        if let Some(handle) = self.thread.take() {
+            if let Err(e) = handle.join() {
+                log::warn!("Progress bar thread did not exit cleanly: {:?}", e);
+            }
+        }
+    }
+
+    /// Update the state of the status bar.
+    pub fn mutate_state(&self, f: impl FnOnce(&mut T)) {
+        f(&mut *self.state.write().unwrap())
+    }
+}
+
+pub struct ShadowStatusBarState {
+    start: std::time::Instant,
+    current: SimulationTime,
```

Comment:
Let's avoid using SimulationTime in new Rust code, since that type is ambiguous. These can both be EmulationTime. In the Display attribute we can still display the elapsed time as time elapsed since sim start using something like `current.time_since(emulated_time.SIMULATION_START)`

Reply:
Thanks, I forgot about this! It's using an `EmulatedTime` internally now.